src/coreclr/jit/objectalloc.cpp

```diff
@@ -815,6 +930,96 @@ void ObjectAllocator::RewriteUses()
 
             return Compiler::fgWalkResult::WALK_CONTINUE;
         }
+
+        Compiler::fgWalkResult PostOrderVisit(GenTree** use, GenTree* user)
+        {
+            GenTree* const tree = *use;
+
+            // Remove GT_BOX, if stack allocated
+            //
+            if (tree->OperIs(GT_BOX))
+            {
+                GenTree* const boxLcl = tree->AsOp()->gtGetOp1();
+                assert(boxLcl->OperIs(GT_LCL_VAR, GT_LCL_ADDR));
+                if (boxLcl->OperIs(GT_LCL_ADDR))
+                {
+                    JITDUMP("Removing BOX wrapper [%06u]\n", m_compiler->dspTreeID(tree));
+                    *use = boxLcl;
+                }
+            }
+            // Make box accesses explicit for UNBOX_HELPER
+            //
+            else if (tree->IsCall())
+            {
+                GenTreeCall* const call = tree->AsCall();
+
+                if (call->IsHelperCall(m_compiler, CORINFO_HELP_UNBOX))
+                {
+                    JITDUMP("Found unbox helper call [%06u]\n", m_compiler->dspTreeID(call));
+
+                    // See if first arg is possibly a stack allocated box or ref class
+                    // (arg will have been retyped local or local address)
+                    //
+                    CallArg*       secondArg     = call->gtArgs.GetArgByIndex(1);
+                    GenTree* const secondArgNode = secondArg->GetNode();
+
+                    if ((secondArgNode->OperIsLocal() || secondArgNode->OperIs(GT_LCL_ADDR)) &&
+                        !secondArgNode->TypeIs(TYP_REF))
```

Comment:
What guarantees that `secondArgNode` in the `secondArgNode->OperIsLocal()` case is pointing to something stack allocated?

Reply:
Just the rewritten type. I can validate the local num too I suppose.

Reply:
Actually not quite this simple, I would need to track all locals that get rewritten. I think checking the type is ok.

We can safely do this transformation for all unbox helper calls, but if the box is on the heap it will just increase code size with no benefit.

Reply:
It seems like doing this transformation during local morph would make its status as "avoiding address exposure" a bit more clear, but I think this is ok then.